lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -0,0 +1,269 @@
+	cidrname := names.CIDRToName(k.(model.BlockAffinityKey).CIDR)
+
+	// Include the hostname as well.
+	host = k.(model.BlockAffinityKey).Host
+	name = fmt.Sprintf("%s-%s", host, cidrname)
+
+	if len(name) >= 253 {
+		// If the name is too long, we need to shorten it.
+		// Remove enough characters to get it below the 253 character limit,
+		// as well as 11 characters to add a hash which helps with uniqueness,
+		// and two characters for the `-` separators between clauses.
+		name = fmt.Sprintf("%s-%s", host[:252-len(cidrname)-13], cidrname)
+
+		// Add a hash to help with uniqueness.
+		// Kubernetes requires all names to end with an alphabetic character, so
+		// append a 'c' to the end to ensure we always meet this requirement.
+		h := sha256.New()
+		h.Write([]byte(fmt.Sprintf("%s+%s", host, cidrname)))
+		name = fmt.Sprintf("%s-%sc", name, hex.EncodeToString(h.Sum(nil))[:10])
+	}
+	return
+}
+
+// toV3 takes the given v1 KVPair and converts it into a v3 representation, suitable
+// for writing as a CRD to the Kubernetes API.
+func (c affinityBlockClient) toV3(kvpv1 *model.KVPair) *model.KVPair {
+	name, cidr, host := c.parseKey(kvpv1.Key)
+	state := kvpv1.Value.(*model.BlockAffinity).State
+	return &model.KVPair{
+		Key: model.ResourceKey{
+			Name: name,
+			Kind: apiv3.KindBlockAffinity,
+		},
+		Value: &apiv3.BlockAffinity{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       apiv3.KindBlockAffinity,
+				APIVersion: "crd.projectcalico.org/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				ResourceVersion: kvpv1.Revision,
+			},
+			Spec: apiv3.BlockAffinitySpec{
+				State: string(state),
+				Node:  host,
+				CIDR:  cidr,
+			},
+		},
+		Revision: kvpv1.Revision,
+	}
+}
+
+func (c *affinityBlockClient) Create(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	nkvp := c.toV3(kvp)
+	kvp, err := c.rc.Create(ctx, nkvp)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) Update(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	nkvp := c.toV3(kvp)
+	kvp, err := c.rc.Update(ctx, nkvp)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) DeleteKVP(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	name, _, _ := c.parseKey(kvp.Key)
+	k := model.ResourceKey{
+		Name: name,
+		Kind: apiv3.KindBlockAffinity,
+	}
+	kvp, err := c.rc.Delete(ctx, k, kvp.Revision, kvp.UID)
+	if err != nil {
+		return nil, err
+	}
+	v1kvp, err := c.toV1(kvp)
+	if err != nil {
+		return nil, err
+	}
+	return v1kvp, nil
+}
+
+func (c *affinityBlockClient) Delete(ctx context.Context, key model.Key, revision string, uid *types.UID) (*model.KVPair, error) {
+	log.Warn("Operation Delete is not supported on BlockAffinity type - use DeleteKVP")
```

Comment:
Worth a comment saying `For correctness, deleting a block requires the UID.`.  Should avoid someone adding int he "missing" method during maintenance.